File: Reasoning-wrapper-master/test_layers.js
const blueprint = [
  {name: "Planner", inputs: []},
  {name: "Subagent 1", inputs: [0]},
  {name: "Subagent 2", inputs: [0]},
  {name: "Subagent 3", inputs: [0]},
  {name: "Subagent 4", inputs: [0]},
  {name: "Writer", inputs: []}
];

const blueprintLayers = {};
blueprint.forEach(node => blueprintLayers[node.name] = 0);

let changed = true;
for (let step = 0; step < blueprint.length + 5 && changed; step++) {
  changed = false;
  blueprint.forEach(node => {
    let maxParentLayer = -1;
    let parentIndices = node.inputs || [];
    if (parentIndices.length === 0 && blueprint.findIndex(n => n.name === node.name) > 0) {
      parentIndices = [blueprint.findIndex(n => n.name === node.name) - 1];
    }
    parentIndices.forEach(parentIdx => {
      if (parentIdx >= 0 && parentIdx < blueprint.length) {
        const parentName = blueprint[parentIdx].name;
        if (parentName in blueprintLayers) {
          maxParentLayer = Math.max(maxParentLayer, blueprintLayers[parentName] ?? -1);
        }
      }
    });
    if (maxParentLayer !== -1 && blueprintLayers[node.name] !== maxParentLayer + 1) {
      blueprintLayers[node.name] = maxParentLayer + 1;
      changed = true;
    }
  });
}

console.log(blueprintLayers);

File: Reasoning-wrapper-master/test_layers.py
blueprint = [
  {"name": "Planner", "inputs": []},
  {"name": "Subagent 1", "inputs": [0]},
  {"name": "Subagent 2", "inputs": [0]},
  {"name": "Subagent 3", "inputs": [0]},
  {"name": "Subagent 4", "inputs": [0]},
  {"name": "Writer", "inputs": []}
]

blueprintLayers = {node["name"]: 0 for node in blueprint}

changed = True
while changed:
    changed = False
    for i, node in enumerate(blueprint):
        maxParentLayer = -1
        parentIndices = node.get("inputs") or []
        if len(parentIndices) == 0 and i > 0:
            parentIndices = [i - 1]
            
        for parentIdx in parentIndices:
            if 0 <= parentIdx < len(blueprint):
                parentName = blueprint[parentIdx]["name"]
                if parentName in blueprintLayers:
                    maxParentLayer = max(maxParentLayer, blueprintLayers.get(parentName, -1))
        
        if maxParentLayer != -1 and blueprintLayers[node["name"]] != maxParentLayer + 1:
            blueprintLayers[node["name"]] = maxParentLayer + 1
            changed = True

print(blueprintLayers)
